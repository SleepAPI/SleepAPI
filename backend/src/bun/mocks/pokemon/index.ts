export * from './mock-pokemon-with-ingredients.js';
