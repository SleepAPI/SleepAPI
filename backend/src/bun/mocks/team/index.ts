export * from './mock-cooking-result.js';
export * from './mock-member-production.js';
export * from './mock-simple-team-results.js';
export * from './mock-team-member-ext.js';
export * from './mock-team-member-settings-ext.js';
export * from './mock-team-result.js';
export * from './mock-team-settings.js';
