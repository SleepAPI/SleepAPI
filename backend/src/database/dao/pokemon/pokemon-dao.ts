import { Static, Type } from '@sinclair/typebox';
import { AbstractDAO, DBWithVersionedIdSchema } from '@src/database/dao/abstract-dao';
import { IngredientError } from '@src/domain/error/ingredient/ingredient-error';
import { IngredientInstance, SubskillInstance } from 'sleepapi-common';

const DBPokemonSchema = Type.Composite([
  DBWithVersionedIdSchema,
  Type.Object({
    fk_user_id: Type.Number(),
    saved: Type.Boolean(),
    shiny: Type.Boolean(),
    gender: Type.Optional(Type.Union([Type.Literal('male'), Type.Literal('female')])),
    external_id: Type.String({ minLength: 36, maxLength: 36 }),
    pokemon: Type.String(),
    name: Type.String(),
    skill_level: Type.Number(),
    carry_size: Type.Number(),
    level: Type.Number(),
    ribbon: Type.Number(),
    nature: Type.String(),
    subskill_10: Type.Optional(Type.String()),
    subskill_25: Type.Optional(Type.String()),
    subskill_50: Type.Optional(Type.String()),
    subskill_75: Type.Optional(Type.String()),
    subskill_100: Type.Optional(Type.String()),
    ingredient_0: Type.String(),
    ingredient_30: Type.String(),
    ingredient_60: Type.String(),
  }),
]);
export type DBPokemon = Static<typeof DBPokemonSchema>;
export type DBPokemonWithoutVersion = Omit<DBPokemon, 'id' | 'version'>;

class PokemonDAOImpl extends AbstractDAO<typeof DBPokemonSchema> {
  public tableName = 'pokemon';
  public schema = DBPokemonSchema;

  public filterFilledSubskills(subskills: DBPokemon): SubskillInstance[] {
    const filledSubskills: SubskillInstance[] = [];

    if (subskills.subskill_10) {
      filledSubskills.push({ level: 10, subskill: subskills.subskill_10 });
    }
    if (subskills.subskill_25) {
      filledSubskills.push({ level: 25, subskill: subskills.subskill_25 });
    }
    if (subskills.subskill_50) {
      filledSubskills.push({ level: 50, subskill: subskills.subskill_50 });
    }
    if (subskills.subskill_75) {
      filledSubskills.push({ level: 75, subskill: subskills.subskill_75 });
    }
    if (subskills.subskill_100) {
      filledSubskills.push({ level: 100, subskill: subskills.subskill_100 });
    }

    return filledSubskills;
  }

  public subskillForLevel(level: number, subskills: SubskillInstance[]) {
    return subskills.find((subskill) => subskill.level === level)?.subskill;
  }

  public ingredientForLevel(level: number, ingredients: IngredientInstance[]) {
    const ingredient = ingredients.find((ingredient) => ingredient.level === level)?.ingredient;
    if (!ingredient) {
      throw new IngredientError('Missing required ingredient in upsert member request for level: ' + level);
    }
    return ingredient;
  }
}

export const PokemonDAO = new PokemonDAOImpl();
