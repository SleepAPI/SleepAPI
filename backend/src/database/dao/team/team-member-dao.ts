import type { Static } from '@sinclair/typebox';
import { Type } from '@sinclair/typebox';
import { AbstractDAO, DBWithVersionedIdSchema } from '@src/database/dao/abstract-dao';

const DBTeamMemberSchema = Type.Composite([
  DBWithVersionedIdSchema,
  Type.Object({
    fk_team_id: Type.Number(),
    fk_pokemon_id: Type.Number(),
    member_index: Type.Number()
  })
]);
export type DBTeamMember = Static<typeof DBTeamMemberSchema>;

class TeamMemberDAOImpl extends AbstractDAO<typeof DBTeamMemberSchema> {
  public tableName = 'team_member';
  public schema = DBTeamMemberSchema;
}

export const TeamMemberDAO = new TeamMemberDAOImpl();
