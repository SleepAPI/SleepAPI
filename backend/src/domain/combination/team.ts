import { PokemonIngredientSet, Time, nature, subskill } from 'sleepapi-common';

export interface TeamSettings {
  camp: boolean;
  bedtime: Time;
  wakeup: Time;
}

export interface TeamMember {
  pokemonSet: PokemonIngredientSet;
  level: number;
  ribbon: number;
  carrySize: number;
  skillLevel: number;
  nature: nature.Nature;
  subskills: subskill.SubSkill[];
  externalId?: string;
}
