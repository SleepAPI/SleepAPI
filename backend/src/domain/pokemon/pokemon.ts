/**
 * Base ing% from Mathcord RP data project: https://docs.google.com/spreadsheets/d/1kBrPl0pdAO8gjOf_NrTgAPseFtqQA27fdfEbMBBeAhs/edit?usp=sharing
 * Base skill% from Raenonx: https://pks.raenonx.cc/en/info/production
 *
 * Skill% is very experimental, and even Raenon seems to change values with big jumps frequently, lots of Pok√©mon also have no skill% on Raenon
 * Mathcord RP project has not published any skill% values yet
 * Skill% is currently not used in Sleep API for this reason
 */

import { Berry } from '../produce/berry';
import { IngredientDrop } from '../produce/ingredient';
import { MainSkill } from '../stat/mainskill';
import { BERRY_SPECIALISTS } from './berry-pokemon';
import { INGREDIENT_SPECIALISTS } from './ingredient-pokemon';
import { SKILL_SPECIALISTS } from './skill-pokemon';

export interface Pokemon {
  name: string;
  specialty: 'berry' | 'ingredient' | 'skill';
  frequency: number;
  ingredientPercentage: number;
  skillPercentage: number;
  berry: Berry;
  carrySize: number;
  maxCarrySize: number;
  ingredient0: IngredientDrop;
  ingredient30: IngredientDrop[];
  ingredient60: IngredientDrop[];
  skill: MainSkill;
}

export const POKEDEX: Pokemon[] = [...BERRY_SPECIALISTS, ...INGREDIENT_SPECIALISTS, ...SKILL_SPECIALISTS];
