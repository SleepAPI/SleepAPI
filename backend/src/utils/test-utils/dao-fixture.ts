import knex, { Knex } from 'knex';
import { DatabaseService } from '../../database/database-service';
import DatabaseMigration from '../../database/migration/database-migration';
import { MockService } from './mock-service';

type InitParams = {
  enforceForeignKeyConstraints?: boolean;
  recreateDatabasesBeforeEachTest?: boolean;
};

async function truncateAllTables(params?: InitParams, knex?: Knex): Promise<void> {
  if (!knex) {
    return;
  }

  const SQLITE_SEQUENCE_TABLE = 'sqlite_sequence';
  const EXCLUDE_PREFIX = ['knex', 'sqlite'];

  const tables = await knex('sqlite_master').where('type', 'table');
  const tableNames = tables
    ?.map((row) => row.name)
    .filter((tableName) => !EXCLUDE_PREFIX.some((prefix) => tableName.startsWith(prefix)));

  await knex.raw('PRAGMA foreign_keys = OFF');

  await Promise.all(tableNames?.map((tableName) => knex(tableName).truncate()) ?? []);
  if (await knex.schema.hasTable(SQLITE_SEQUENCE_TABLE)) {
    await knex(SQLITE_SEQUENCE_TABLE).truncate();
  }

  if (params?.enforceForeignKeyConstraints) {
    await knex?.raw('PRAGMA foreign_keys = ON');
  }
}

export const DaoFixture = {
  init(params?: InitParams) {
    let pokemonsleepDB: Knex;

    if (params?.recreateDatabasesBeforeEachTest) {
      beforeEach(async () => {
        await setup();
      });
    } else {
      beforeAll(async () => {
        await setup();
      });
    }

    async function setup() {
      MockService.init({ DatabaseService });

      pokemonsleepDB = knex({
        client: 'sqlite3',
        useNullAsDefault: true,
        connection: ':memory:',
        acquireConnectionTimeout: 10000,
      });

      Object.defineProperty(DatabaseService, '#knex', {
        configurable: true,
        value: pokemonsleepDB,
      });

      if (params?.enforceForeignKeyConstraints) {
        await pokemonsleepDB?.raw('PRAGMA foreign_keys = ON');
      }

      DatabaseService.getKnex = () => Promise.resolve(pokemonsleepDB);
      await DatabaseMigration.migrate();
    }

    async function destroyDatabases() {
      await pokemonsleepDB.destroy();
    }

    afterEach(async () => {
      if (params?.recreateDatabasesBeforeEachTest) {
        await destroyDatabases();
      } else {
        await truncateAllTables(params, pokemonsleepDB);
      }

      MockService.restore();
    });

    afterAll(async () => {
      if (!params?.recreateDatabasesBeforeEachTest) {
        await destroyDatabases();
      }
    });
  },
};
