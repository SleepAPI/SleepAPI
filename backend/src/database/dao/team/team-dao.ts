import { Static, Type } from '@sinclair/typebox';
import { AbstractDAO, DBWithVersionedIdSchema } from '@src/database/dao/abstract-dao';
import { DBPokemon, PokemonDAO } from '@src/database/dao/pokemon/pokemon-dao';
import { TeamMemberDAO } from '@src/database/dao/team/team-member-dao';
import { GetTeamResponse, MemberInstance, SubskillInstance } from 'sleepapi-common';

const DBTeamSchema = Type.Composite([
  DBWithVersionedIdSchema,
  Type.Object({
    fk_user_id: Type.Number(),
    team_index: Type.Number(),
    name: Type.String(),
    camp: Type.Boolean(),
    bedtime: Type.String(),
    wakeup: Type.String(),
  }),
]);
export type DBTeam = Static<typeof DBTeamSchema>;
export type DBTeamWithoutVersion = Omit<DBTeam, 'id' | 'version'>;

class TeamDAOImpl extends AbstractDAO<typeof DBTeamSchema> {
  public tableName = 'team';
  public schema = DBTeamSchema;

  public async findTeamsWithMembers(userId: number) {
    const teams = await this.findMultiple({ fk_user_id: userId });

    const teamsWithMembers: GetTeamResponse[] = [];
    for (const team of teams) {
      const members: MemberInstance[] = [];
      const memberMetaData = await TeamMemberDAO.findMultiple({ fk_team_id: team.id });
      for (const memberData of memberMetaData) {
        const member: DBPokemon = await PokemonDAO.get({ id: memberData.fk_pokemon_id });

        const subskills: SubskillInstance[] = [];
        if (member.subskill_10) {
          subskills.push({ level: 10, subskill: member.subskill_10 });
        }
        if (member.subskill_25) {
          subskills.push({ level: 25, subskill: member.subskill_25 });
        }
        if (member.subskill_50) {
          subskills.push({ level: 50, subskill: member.subskill_50 });
        }
        if (member.subskill_75) {
          subskills.push({ level: 75, subskill: member.subskill_75 });
        }
        if (member.subskill_100) {
          subskills.push({ level: 100, subskill: member.subskill_100 });
        }

        members.push({
          memberIndex: memberData.member_index,
          version: member.version,
          saved: member.saved,
          shiny: member.shiny,
          externalId: member.external_id,
          pokemon: member.pokemon,
          name: member.name,
          level: member.level,
          ribbon: member.ribbon,
          carrySize: member.carry_size,
          skillLevel: member.skill_level,
          nature: member.nature,
          subskills,
          ingredients: [
            {
              level: 0,
              ingredient: member.ingredient_0,
            },
            {
              level: 30,
              ingredient: member.ingredient_30,
            },
            {
              level: 60,
              ingredient: member.ingredient_60,
            },
          ],
        });
      }

      teamsWithMembers.push({
        index: team.team_index,
        name: team.name,
        camp: team.camp,
        bedtime: team.bedtime,
        wakeup: team.wakeup,
        version: team.version,
        members,
      });
    }
    return teamsWithMembers;
  }

  // TODO: if we get heavy load on server we can do it on db, but needs some boolean processing and more love
  // public async getTeamsWithPokemon(userId: number): Promise<GetTeamResponse[]> {
  //   const knex = await DatabaseService.getKnex();

  //   const rows = await knex
  //     .select([
  //       'team.id as team_id',
  //       'team.version as team_version',
  //       'team.team_index',
  //       'team.name as team_name',
  //       'team.camp',
  //       'team_member.id as member_id',
  //       'team_member.member_index',
  //       'pokemon.id as pokemon_id',
  //       'pokemon.saved',
  //       'pokemon.pokemon',
  //       'pokemon.name as pokemon_name',
  //       'pokemon.level',
  //       'pokemon.carry_size',
  //       'pokemon.skill_level',
  //       'pokemon.nature',
  //       'pokemon.subskill_10',
  //       'pokemon.subskill_25',
  //       'pokemon.subskill_50',
  //       'pokemon.subskill_75',
  //       'pokemon.subskill_100',
  //       'pokemon.ingredient_0',
  //       'pokemon.ingredient_30',
  //       'pokemon.ingredient_60',
  //     ])
  //     .from('team')
  //     .leftJoin('team_member', 'team.id', 'team_member.fk_team_id')
  //     .leftJoin('pokemon', 'team_member.fk_pokemon_id', 'pokemon.id')
  //     .where('team.fk_user_id', userId)
  //     .orderBy(['team.team_index', 'team_member.member_index']);

  //   const teamsMap: Record<number, GetTeamResponse> = {};

  //   for (const row of rows) {
  //     if (!teamsMap[row.team_id]) {
  //       teamsMap[row.team_id] = {
  //         index: row.team_index,
  //         name: row.team_name,
  //         camp: row.camp,
  //         version: row.team_version,
  //         members: [],
  //       };
  //     }

  //     if (row.pokemon_id !== null) {
  //       const subskills: InstancedSubskill[] = [];
  //       if (row.subskill_10) {
  //         subskills.push({ level: 10, subskill: row.subskill_10 });
  //       }
  //       if (row.subskill_25) {
  //         subskills.push({ level: 25, subskill: row.subskill_25 });
  //       }
  //       if (row.subskill_50) {
  //         subskills.push({ level: 50, subskill: row.subskill_50 });
  //       }
  //       if (row.subskill_75) {
  //         subskills.push({ level: 75, subskill: row.subskill_75 });
  //       }
  //       if (row.subskill_100) {
  //         subskills.push({ level: 100, subskill: row.subskill_100 });
  //       }

  //       teamsMap[row.team_id].members[row.member_index] = {
  //         index: row.member_index,
  //         saved: row.saved,
  //         pokemon: row.pokemon,
  //         name: row.pokemon_name,
  //         level: row.level,
  //         carrySize: row.carry_size,
  //         skillLevel: row.skill_level,
  //         nature: row.nature,
  //         subskills,
  //         ingredients: [
  //           {
  //             level: 0,
  //             ingredient: row.ingredient_0,
  //           },
  //           {
  //             level: 30,
  //             ingredient: row.ingredient_30,
  //           },
  //           {
  //             level: 60,
  //             ingredient: row.ingredient_60,
  //           },
  //         ],
  //       };
  //     }
  //   }

  //   return Object.values(teamsMap);
  // }
}

export const TeamDAO = new TeamDAOImpl();
