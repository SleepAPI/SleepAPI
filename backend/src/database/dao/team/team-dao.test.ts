/* eslint-disable @typescript-eslint/no-explicit-any */
import { PokemonDAO } from '@src/database/dao/pokemon/pokemon-dao';
import { TeamDAO } from '@src/database/dao/team/team-dao';
import { TeamMemberDAO } from '@src/database/dao/team/team-member-dao';
import { DaoFixture } from '@src/utils/test-utils/dao-fixture';
import { uuid } from 'sleepapi-common';

DaoFixture.init({ recreateDatabasesBeforeEachTest: true });

beforeEach(() => {
  let innerCounter = 0;
  uuid.v4 = jest.fn(() => `${++innerCounter}`.padEnd(36, `${innerCounter}`));
});

describe('TeamDAO insert', () => {
  it('shall insert new entity', async () => {
    const team = await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 0,
      name: 'Team A',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });
    expect(team).toBeDefined();

    const data = await TeamDAO.findMultiple();
    expect(data).toMatchInlineSnapshot(`
      [
        {
          "bedtime": "21:30",
          "camp": false,
          "fk_user_id": 1,
          "id": 1,
          "name": "Team A",
          "team_index": 0,
          "version": 1,
          "wakeup": "06:00",
        },
      ]
    `);
  });

  it('shall fail to insert entity without fk_user_id', async () => {
    await expect(
      TeamDAO.insert({
        fk_user_id: undefined as any,
        team_index: 0,
        name: 'Team A',
        camp: false,
        bedtime: '21:30',
        wakeup: '06:00',
      })
    ).rejects.toThrow(/SQLITE_CONSTRAINT: NOT NULL constraint failed: team.fk_user_id/);
  });

  it('shall fail to insert entity without team_index', async () => {
    await expect(
      TeamDAO.insert({
        fk_user_id: 1,
        team_index: undefined as any,
        name: 'Team A',
        camp: false,
        bedtime: '21:30',
        wakeup: '06:00',
      })
    ).rejects.toThrow(/SQLITE_CONSTRAINT: NOT NULL constraint failed: team.team_index/);
  });

  it('shall fail to insert entity without name', async () => {
    await expect(
      TeamDAO.insert({
        fk_user_id: 1,
        team_index: 0,
        name: undefined as any,
        camp: false,
        bedtime: '21:30',
        wakeup: '06:00',
      })
    ).rejects.toThrow(/SQLITE_CONSTRAINT: NOT NULL constraint failed: team.name/);
  });

  it('shall fail to insert entity with duplicate fk_user_id and team_index', async () => {
    await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 0,
      name: 'Team A',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });
    await expect(
      TeamDAO.insert({
        fk_user_id: 1,
        team_index: 0,
        name: 'Team B',
        camp: true,
        bedtime: '21:30',
        wakeup: '06:00',
      })
    ).rejects.toThrow(/SQLITE_CONSTRAINT: UNIQUE constraint failed: team.fk_user_id, team.team_index/);
  });
});

describe('TeamDAO update', () => {
  it('shall update entity', async () => {
    const team = await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 0,
      name: 'Team A',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });
    expect(team.name).toEqual('Team A');

    await TeamDAO.update({ ...team, name: 'Updated Team A' });

    const data = await TeamDAO.findMultiple();
    expect(data).toMatchInlineSnapshot(`
      [
        {
          "bedtime": "21:30",
          "camp": false,
          "fk_user_id": 1,
          "id": 1,
          "name": "Updated Team A",
          "team_index": 0,
          "version": 2,
          "wakeup": "06:00",
        },
      ]
    `);
  });

  it('shall fail to update entity with duplicate fk_user_id and team_index', async () => {
    await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 0,
      name: 'Team A',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });
    const teamB = await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 1,
      name: 'Team B',
      camp: true,
      bedtime: '21:30',
      wakeup: '06:00',
    });

    await expect(TeamDAO.update({ ...teamB, team_index: 0 })).rejects.toThrow(
      /SQLITE_CONSTRAINT: UNIQUE constraint failed: team.fk_user_id, team.team_index/
    );
  });
});

describe('TeamDAO delete', () => {
  it('shall delete entity', async () => {
    const team = await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 0,
      name: 'Team A',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });

    await TeamDAO.delete({ id: team.id });

    const data = await TeamDAO.findMultiple();
    expect(data).toEqual([]);
  });
});

describe('findTeamsWithMembers', () => {
  it('shall get team with members', async () => {
    await TeamDAO.insert({
      fk_user_id: 1,
      team_index: 0,
      name: 'Team A',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });
    // should not be found
    await TeamDAO.insert({
      fk_user_id: 2,
      team_index: 0,
      name: 'Team B',
      camp: false,
      bedtime: '21:30',
      wakeup: '06:00',
    });

    await PokemonDAO.insert({
      fk_user_id: 1,
      saved: true,
      shiny: false,
      external_id: uuid.v4(),
      pokemon: 'Pikachu',
      name: 'Sparky',
      skill_level: 5,
      carry_size: 10,
      level: 25,
      ribbon: 0,
      nature: 'Brave',
      subskill_10: 'Thunderbolt',
      subskill_25: 'Quick Attack',
      subskill_50: 'Iron Tail',
      subskill_75: 'Electro Ball',
      subskill_100: 'Thunder',
      ingredient_0: 'Berry',
      ingredient_30: 'Potion',
      ingredient_60: 'Elixir',
    });

    await PokemonDAO.insert({
      fk_user_id: 1,
      saved: true,
      shiny: false,
      external_id: uuid.v4(),
      pokemon: 'Bulbasaur',
      name: 'Bubble',
      skill_level: 5,
      carry_size: 10,
      level: 25,
      ribbon: 0,
      nature: 'Brave',
      subskill_10: 'Thunderbolt',
      subskill_25: 'Quick Attack',
      subskill_50: 'Iron Tail',
      subskill_75: 'Electro Ball',
      subskill_100: 'Thunder',
      ingredient_0: 'Berry',
      ingredient_30: 'Potion',
      ingredient_60: 'Elixir',
    });

    // should not be included, is saved to user, but not included in team
    await PokemonDAO.insert({
      fk_user_id: 1,
      saved: true,
      shiny: false,
      external_id: uuid.v4(),
      pokemon: 'Bulbasaur',
      name: 'Bubble',
      skill_level: 5,
      carry_size: 10,
      level: 25,
      ribbon: 0,
      nature: 'Brave',
      subskill_10: 'Thunderbolt',
      subskill_25: 'Quick Attack',
      subskill_50: 'Iron Tail',
      subskill_75: 'Electro Ball',
      subskill_100: 'Thunder',
      ingredient_0: 'Berry',
      ingredient_30: 'Potion',
      ingredient_60: 'Elixir',
    });

    // Team: [pika, pika, bulba]
    await TeamMemberDAO.insert({
      member_index: 0,
      fk_team_id: 1,
      fk_pokemon_id: 1,
    });
    await TeamMemberDAO.insert({
      member_index: 1,
      fk_team_id: 1,
      fk_pokemon_id: 1,
    });
    await TeamMemberDAO.insert({
      member_index: 2,
      fk_team_id: 1,
      fk_pokemon_id: 2,
    });

    expect(await TeamDAO.findTeamsWithMembers(1)).toMatchSnapshot();
  });
});
