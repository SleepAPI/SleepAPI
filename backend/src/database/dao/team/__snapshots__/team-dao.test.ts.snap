// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`findTeamsWithMembers shall get team with members 1`] = `
[
  {
    "bedtime": "21:30",
    "camp": false,
    "favoredBerries": undefined,
    "index": 0,
    "members": [
      {
        "carrySize": 10,
        "externalId": "111111111111111111111111111111111111",
        "gender": undefined,
        "ingredients": [
          {
            "ingredient": "Berry",
            "level": 0,
          },
          {
            "ingredient": "Potion",
            "level": 30,
          },
          {
            "ingredient": "Elixir",
            "level": 60,
          },
        ],
        "level": 25,
        "memberIndex": 0,
        "name": "Sparky",
        "nature": "Brave",
        "pokemon": "Pikachu",
        "ribbon": 0,
        "saved": true,
        "shiny": false,
        "skillLevel": 5,
        "subskills": [
          {
            "level": 10,
            "subskill": "Thunderbolt",
          },
          {
            "level": 25,
            "subskill": "Quick Attack",
          },
          {
            "level": 50,
            "subskill": "Iron Tail",
          },
          {
            "level": 75,
            "subskill": "Electro Ball",
          },
          {
            "level": 100,
            "subskill": "Thunder",
          },
        ],
        "version": 1,
      },
      {
        "carrySize": 10,
        "externalId": "111111111111111111111111111111111111",
        "gender": undefined,
        "ingredients": [
          {
            "ingredient": "Berry",
            "level": 0,
          },
          {
            "ingredient": "Potion",
            "level": 30,
          },
          {
            "ingredient": "Elixir",
            "level": 60,
          },
        ],
        "level": 25,
        "memberIndex": 1,
        "name": "Sparky",
        "nature": "Brave",
        "pokemon": "Pikachu",
        "ribbon": 0,
        "saved": true,
        "shiny": false,
        "skillLevel": 5,
        "subskills": [
          {
            "level": 10,
            "subskill": "Thunderbolt",
          },
          {
            "level": 25,
            "subskill": "Quick Attack",
          },
          {
            "level": 50,
            "subskill": "Iron Tail",
          },
          {
            "level": 75,
            "subskill": "Electro Ball",
          },
          {
            "level": 100,
            "subskill": "Thunder",
          },
        ],
        "version": 1,
      },
      {
        "carrySize": 10,
        "externalId": "222222222222222222222222222222222222",
        "gender": undefined,
        "ingredients": [
          {
            "ingredient": "Berry",
            "level": 0,
          },
          {
            "ingredient": "Potion",
            "level": 30,
          },
          {
            "ingredient": "Elixir",
            "level": 60,
          },
        ],
        "level": 25,
        "memberIndex": 2,
        "name": "Bubble",
        "nature": "Brave",
        "pokemon": "Bulbasaur",
        "ribbon": 0,
        "saved": true,
        "shiny": false,
        "skillLevel": 5,
        "subskills": [
          {
            "level": 10,
            "subskill": "Thunderbolt",
          },
          {
            "level": 25,
            "subskill": "Quick Attack",
          },
          {
            "level": 50,
            "subskill": "Iron Tail",
          },
          {
            "level": 75,
            "subskill": "Electro Ball",
          },
          {
            "level": 100,
            "subskill": "Thunder",
          },
        ],
        "version": 1,
      },
    ],
    "name": "Team A",
    "recipeType": "curry",
    "version": 1,
    "wakeup": "06:00",
  },
]
`;
